integration/run_test.go

```diff
@@ -0,0 +1,128 @@
+*/
+
+package integration
+
+import (
+	"fmt"
+	"regexp"
+	"testing"
+	"time"
+
+	"github.com/awslabs/soci-snapshotter/util/testutil"
+)
+
+// TestRunMultipleContainersWithSameImage runs two containers with the same image at the same time
+func TestRunMultipleContainersWithSameImage(t *testing.T) {
+	RunMultipleWebContainers(t, []string{nginxImage, nginxImage})
+}
+
+// TestRunMultipleContainersWithDifferentImages runs two containers with different images at the same time
+func TestRunMultipleContainersWithDifferentImages(t *testing.T) {
+	RunMultipleWebContainers(t, []string{nginxImage, drupalImage})
+}
+
+func RunMultipleWebContainers(t *testing.T, imageNames []string) {
+	regConfig := newRegistryConfig()
+	sh, done := newShellWithRegistry(t, regConfig)
+	defer done()
+
+	// Initialize config files for containerd and snapshotter
+	additionalConfig := ""
+	if !isTestingBuiltinSnapshotter() {
+		additionalConfig = proxySnapshotterConfig
+	}
+	containerdConfigYaml := testutil.ApplyTextTemplate(t, `
+version = 2
+
+[plugins."io.containerd.snapshotter.v1.soci"]
+root_path = "/var/lib/soci-snapshotter-grpc/"
+
+[plugins."io.containerd.snapshotter.v1.soci".blob]
+check_always = true
+
+{{.AdditionalConfig}}
+`, struct {
+		AdditionalConfig string
+	}{
+		AdditionalConfig: additionalConfig,
+	})
+
+	// Setup environment
+	if err := testutil.WriteFileContents(sh, defaultContainerdConfigPath, []byte(containerdConfigYaml), 0600); err != nil {
+		t.Fatalf("failed to write %v: %v", defaultContainerdConfigPath, err)
+	}
+	sh.
+		X("update-ca-certificates").
+		Retry(100, "nerdctl", "login", "-u", regConfig.user, "-p", regConfig.pass, regConfig.host)
+
+	rebootContainerd(t, sh, "", "")
+	for _, imageName := range deduplicate(imageNames) {
+		// Mirror image
+		copyImage(sh, dockerhub(imageName), regConfig.mirror(imageName))
+		indexDigest := optimizeImage(sh, regConfig.mirror(imageName))
+
+		// Pull image
+		sh.X("ctr", "i", "pull", "--user", regConfig.creds(), regConfig.mirror(imageName).ref)
+
+		// Create SOCI metadata
+		sh.X("soci", "create", regConfig.mirror(imageName).ref)
+		sh.X("soci", "image", "rpull", "--user", regConfig.creds(), "--soci-index-digest", indexDigest, regConfig.mirror(imageName).ref)
+	}
+
+	// Run the containers
+	for index, imageName := range imageNames {
+		image := regConfig.mirror(imageName).ref
+		containerName := "test_" + fmt.Sprint(index) + "_" + regexp.MustCompile(`[^a-zA-Z0-9 ]+`).ReplaceAllString(imageName, "_")
+		sh.X("soci", "run", "-d", "--rm", "--snapshotter=soci", image, containerName)
+	}
+
+	time.Sleep(5 * time.Second) // FIXME find better solution for "connection refused" from `curl` below too soon
+
+	// Do something in each container
+	for index, imageName := range imageNames {
+		containerName := "test_" + fmt.Sprint(index) + "_" + regexp.MustCompile(`[^a-zA-Z0-9 ]+`).ReplaceAllString(imageName, "_")
+		sh.X("ctr", "task", "exec", "--exec-id", "test-curl", containerName, "curl", "http://127.0.0.1")
```

Comment:
Normally, we don't run a process in the snapshotter. We compare the contents of a directory (`/usr`) with the corresponding contents created by the overlayfs snapshotter. Can we do something similar here?

Reply:
I was under the impression that these tests were intended to be about the functioning of the containers, not just about whether they start up. I tried to aim in the middle here, but I expect we could go so far as a test that would perform some operation then check the results within the container.
Nailing this down probably requires a more specific definition of what exactly we're trying to test.

Reply:
The goal of these tests is to make sure we don't accidentally share state between containers in a way that breaks things.

E.g. a previous version of the snapshotter only lazily loaded the first rpulled image (https://github.com/awslabs/soci-snapshotter/pull/77) and we couldn't detect that from our existing tests which only ran 1 task in a clean environment.

I think verifying that the container workload comes up will tell us that the snapshotter works in this case, but we may want additional verifications. At the very least verifying no fallback to overlayfs. We do that in the single task tests (e.g. https://github.com/awslabs/soci-snapshotter/blob/main/integration/pull_test.go#L234-L241) but it might be easier to do it here after we have a like a metric or something on fallback.

Reply:
@Kern-- does that clear this to merge, or do you want this PR to add the mount-counting check, or for this PR to wait for the fallback metric functionality?

Reply:
Yes, we can merge without the mount-checking. Can you create a follow up ticket?